AppDB/appscale/datastore/scripts/transaction_groomer.py

```diff
@@ -0,0 +1,362 @@
+    node_prefix = '/'.join([self._registration_path, self._groomer_id]) + '-'
+
+    # Make sure an older node from this groomer did not remain.
+    self._clean_created_nodes()
+
+    # The groomer must be registered before it can continue working.
+    while True:
+      try:
+        full_path = self._zk_client.create(node_prefix, ephemeral=True,
+                                           sequence=True)
+        self._node = full_path[len(self._registration_path) + 1:]
+        break
+      except ZookeeperError:
+        self._clean_created_nodes()
+        continue
+
+  def _state_listener(self, state):
+    """ Watches for changes to the ZooKeeper connection state. """
+    if state == KazooState.CONNECTED:
+      IOLoop.instance().add_callback(self._register_groomer)
+
+
+class ProjectGroomer(object):
+  """ Cleans up expired transactions for a project. """
+  def __init__(self, project_id, coordinator, zk_client, db_access,
+               thread_pool):
+    """ Creates a new ProjectGroomer.
+
+    Args:
+      project_id: A string specifying a project ID.
+      coordinator: A GroomingCoordinator.
+      zk_client: A KazooClient.
+      db_access: A DatastoreProxy.
+      thread_pool: A ThreadPoolExecutor.
+    """
+    self.project_id = project_id
+
+    self._coordinator = coordinator
+    self._zk_client = zk_client
+    self._tornado_zk = TornadoKazoo(self._zk_client)
+    self._db_access = db_access
+    self._thread_pool = thread_pool
+    self._txids_path = '/appscale/apps/{}/txids'.format(self.project_id)
+    self._batch_resolver = BatchResolver(self.project_id, self._db_access)
+
+    self._stop_event = AsyncEvent()
+    self._stopped_event = AsyncEvent()
+
+    self._zk_client.ensure_path(self._txids_path)
+    IOLoop.current().spawn_callback(self.start)
+
+  @gen.coroutine
+  def start(self):
+    """ Starts the grooming process until the stop event is set. """
+    logger.info('Grooming {}'.format(self.project_id))
+    while True:
+      if self._stop_event.is_set():
+        break
+
+      try:
+        yield self._groom_project()
+      except Exception:
+        # Prevent the grooming loop from stopping if an error is encountered.
+        logger.exception(
+          'Unexpected error while grooming {}'.format(self.project_id))
+        yield gen.sleep(MAX_TX_DURATION)
+
+    self._stopped_event.set()
+
+  @gen.coroutine
+  def stop(self):
+    """ Stops the grooming process. """
+    logger.info('Stopping grooming process for {}'.format(self.project_id))
+    self._stop_event.set()
+    yield self._stopped_event.wait()
+
+  @gen.coroutine
+  def _groom_project(self):
+    """ Runs the grooming process. """
+    index = self._coordinator.index
+    worker_count = self._coordinator.total_workers
+
+    oldest_valid_tx = yield self._fetch_and_clean(index, worker_count)
+
+    # Wait until there's a reasonable chance that some transactions have
+    # timed out.
+    next_timeout_eta = oldest_valid_tx + MAX_TX_DURATION
```

Comment:
so oldest_valid_tx is a time? If so, could you make the variable name clearer, I assumed it held the transaction.

Reply:
Good point. That wasn't very clear.